### PR TITLE
Deprecate the SecureRandom class

### PR DESCRIPTION
**Pull request by pierredup**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Deprecate the `Symfony\Component\Security\Core\Util\SecureRandom` and encourage the usage of the `random_bytes` function instead

### CODE REVIEW

**Comment by stof** [comment]
This implementation is broken: the Symfony codebase still use the deprecated function, meaning it triggers deprecation warnings for everyone (and this is why the testsuite is failing)

**Comment by pierredup** [comment]
@stof What would be the best way to deprecate this class then? Should the dependency on `paragonie/random_compat` be added here already, with all the instances that use the `SecureRandom` class be updated to use the function instead? That way the class and interface can be deprecated with all the internal code using the new function. Or is there a better way to do the deprecation?

**Comment by stof** [comment]
@pierredup yes, this is the way it should be done. A deprecated API should not be used anymore. If you cannot avoid using it, it cannot be a deprecated API.

File: src/Symfony/Component/Security/Core/composer.json [hunk 1]
```diff
@@ -33,7 +33,8 @@
         "symfony/http-foundation": "",
         "symfony/validator": "For using the user password constraint",
         "symfony/expression-language": "For using the expression voter",
-        "ircmaxell/password-compat": "For using the BCrypt password encoder in PHP <5.5"
+        "ircmaxell/password-compat": "For using the BCrypt password encoder in PHP <5.5",
+        "paragonie/random_compat": "For secure number generation"
```
Comment (derrabus) [inline src/Symfony/Component/Security/Core/composer.json]: I think, we should document that this library is only a compatibility layer for PHP 5: `"For secure random number generation in PHP 5.x"`

**Comment by derrabus** [comment]
Status: Needs Work

**Comment by pierredup** [comment]
I think this PR is ready /cc @stof @fabpot 

The travis failures seems unrelated

Status: Needs Review

**Comment by stof** [comment]
I would like to see our testsuite being fixed and then rerunning the builds for this PR, to ensure things work properly though.

**Comment by pierredup** [comment]
Can the build for this be re-run? Or should I do a push to trigger new builds?

**Comment by stof** [comment]
@pierredup there are conflicts preventing the merge, so this needs a rebase anyway. After pushing the rebased version, a new build with run.

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,7 @@ CHANGELOG
    `Symfony\Component\Security\Http\Authentication\SimpleFormAuthenticatorInterface` instead
  * deprecated `Symfony\Component\Security\Core\Util\ClassUtils`, use
    `Symfony\Component\Security\Acl\Util\ClassUtils` instead
+ * deprecated `Symfony\Component\Security\Core\Util\SecureRandom` class in favour of the `random_bytes` function
```
Comment (xabbuh) [inline src/Symfony/Component/Security/CHANGELOG.md]: deprecated the [...]

Comment (xabbuh) [inline src/Symfony/Component/Security/CHANGELOG.md]: in favor of

File: src/Symfony/Component/Security/composer.json [hunk 3]
```diff
@@ -20,7 +20,8 @@
         "symfony/security-acl": "~2.7",
         "symfony/event-dispatcher": "~2.2|~3.0.0",
         "symfony/http-foundation": "~2.1|~3.0.0",
-        "symfony/http-kernel": "~2.4|~3.0.0"
+        "symfony/http-kernel": "~2.4|~3.0.0",
+        "paragonie/random_compat" : "~1.0"
```
Comment (xabbuh) [inline src/Symfony/Component/Security/composer.json]: should be removed from the suggest section then

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/security_csrf.xml [hunk 4]
```diff
@@ -12,7 +12,6 @@
 
     <services>
         <service id="security.csrf.token_generator" class="%security.csrf.token_generator.class%" public="false">
-            <argument type="service" id="security.secure_random" />
         </service>
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/security_csrf.xml]: You can omit the closing tag and close the opening one directly instead.

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_rememberme.xml [hunk 5]
```diff
@@ -46,7 +46,6 @@
                  class="%security.authentication.rememberme.services.persistent.class%"
                  parent="security.authentication.rememberme.services.abstract"
                  abstract="true">
-            <argument type="service" id="security.secure_random" />
         </service>
```
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_rememberme.xml]: same here

File: src/Symfony/Component/Security/Core/composer.json [hunk 6]
```diff
@@ -35,7 +35,8 @@
         "symfony/validator": "For using the user password constraint",
         "symfony/expression-language": "For using the expression voter",
         "symfony/ldap": "For using LDAP integration",
-        "ircmaxell/password-compat": "For using the BCrypt password encoder in PHP <5.5"
+        "ircmaxell/password-compat": "For using the BCrypt password encoder in PHP <5.5",
+        "paragonie/random_compat": "For secure random number generation in PHP 5.x"
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/composer.json]: Why is it suggested here?

Comment (stof) [inline src/Symfony/Component/Security/Core/composer.json]: this should be a requirement

File: src/Symfony/Component/Security/Http/RememberMe/PersistentTokenBasedRememberMeServices.php [hunk 7]
```diff
@@ -32,11 +32,12 @@
 class PersistentTokenBasedRememberMeServices extends AbstractRememberMeServices
 {
     private $tokenProvider;
-    private $secureRandom;
 
     /**
      * Constructor.
      *
+     * Note: The $secureRandom parameter is deprecated since version 2.8 and will be removed in 3.0.
```
Comment (stof) [inline src/Symfony/Component/Security/Http/RememberMe/PersistentTokenBasedRememberMeServices.php]: you should trigger a deprecation warning when passing one then

**Comment by fabpot** [comment]
@pierredup Can you finish this PR in the next few days?

**Comment by pierredup** [comment]
@fabpot done

**Comment by pierredup** [comment]
The tests are a bit weird, it seems it didn't run the tests at all for PHP 5.4 and 5.5. Not sure if you want to re-run the tests?

**Comment by derrabus** [comment]
`paragonie/random_compat` is now a dependency, ensuring that the `random_bytes()` function always exists. In this case, the old `SecureRandom` class contains a lot of dead code:

```php
public function nextBytes($nbBytes)
{
    if (function_exists('random_bytes')) {
        return random_bytes($nbBytes);
    }

    // try OpenSSL
    if ($this->useOpenSsl) {
        // ...
    }
    // ...
```

We will now always take the early exit, so anything below that `if` block will never be executed. That means, that the `SecureRandom` class boils down to this:

```php
final class SecureRandom implements SecureRandomInterface
{
    /**
     * {@inheritdoc}
     */
    public function nextBytes($nbBytes)
    {
        return random_bytes($nbBytes);
    }
}
```

**Comment by pierredup** [comment]
Thanks @derrabus, I have removed the unused logic

**Comment by stof** [comment]
> The tests are a bit weird, it seems it didn't run the tests at all for PHP 5.4 and 5.5. Not sure if you want to re-run the tests?

That's expected. We don't run them on 5.4 and 5.5 anymore for pull requests, because we can only have 5 concurrent jobs per organization on Travis. This allows us to make build run faster.

File: src/Symfony/Component/Security/Core/Util/SecureRandom.php [hunk 8]
```diff
@@ -11,106 +11,23 @@
 
 namespace Symfony\Component\Security\Core\Util;
 
-use Psr\Log\LoggerInterface;
+@trigger_error('The '.__NAMESPACE__.'\SecureRandom class is deprecated since version 2.8 and will be removed in 3.0. Use the random_bytes function instead.', E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Core/Util/SecureRandom.php]: What about parentheses after the function name (same for the changelog)?

File: src/Symfony/Component/Security/Csrf/TokenGenerator/UriSafeTokenGenerator.php [hunk 9]
```diff
@@ -47,7 +41,10 @@ class UriSafeTokenGenerator implements TokenGeneratorInterface
      */
     public function __construct(SecureRandomInterface $random = null, $entropy = 256)
     {
-        $this->random = $random ?: new SecureRandom();
+        if (null !== $random) {
+            @trigger_error('The $random parameter in '.__METHOD__.' is deprecated since version 2.8 and will be removed in 3.0.', E_USER_DEPRECATED);
+        }
```
Comment (xabbuh) [inline src/Symfony/Component/Security/Csrf/TokenGenerator/UriSafeTokenGenerator.php]: We should make sure that we are able to drop the first argument in Symfony 3.0 and only need to pass the entropy. For this, we need to drop the type hint for the `SecureRandomInterface`:

```php
public function __construct($random, $entropy = 256)
{
    if ($random instanceof SecureRandomInterface || func_num_args() === 2) {
        @trigger_error('The '.__METHOD__.' method now requires the entropy to be given as the first argument. The SecureRandomInterface will be removed in 3.0.');

        $this->entropy = $entropy;
    } elseif (func_num_args() === 0) {
        $this->entropy = 256;
    } else {
        $this->entropy = $random;
    }
}
```

**Comment by pierredup** [comment]
Thanks for the info @stof

**Comment by pierredup** [comment]
All comments have been addressed

**Comment by derrabus** [comment]
Status: Reviewed

**Comment by xabbuh** [comment]
@pierredup You also have to update the `SecureRandomTest` class. The `getSecureRandoms()` data provider tries to modify the `useOpenSsl` property which no longer exists.

**Comment by xabbuh** [comment]
Status: Needs work

**Comment by pierredup** [comment]
Thanks @xabbuh. Is it still necessary for the test, since the SecureRandom class is now basically only a wrapper for the `random_bytes` function? Can I remove the test, or do you prefer I just update it?

**Comment by xabbuh** [comment]
Imo you can simply drop the OpenSSL related part of that test.

**Comment by derrabus** [comment]
Status: Reviewed

**Comment by fabpot** [comment]
:+1: ping @symfony/deciders

File: src/Symfony/Component/Security/Csrf/TokenGenerator/UriSafeTokenGenerator.php [hunk 10]
```diff
@@ -40,15 +32,24 @@ class UriSafeTokenGenerator implements TokenGeneratorInterface
     /**
      * Generates URI-safe CSRF tokens.
      *
-     * @param SecureRandomInterface|null $random  The random value generator used for
-     *                                            generating entropy
-     * @param int                        $entropy The amount of entropy collected for
-     *                                            each token (in bits)
+     * Note: The $random parameter is deprecated since version 2.8 and will be removed in 3.0.
+     *
+     * @param SecureRandomInterface|int|null $random  The random value generator used for
```
Comment (Tobion) [inline src/Symfony/Component/Security/Csrf/TokenGenerator/UriSafeTokenGenerator.php]: the deprecated parameter value should not be documented anymore

File: src/Symfony/Component/Security/Csrf/TokenGenerator/UriSafeTokenGenerator.php [hunk 11]
```diff
@@ -40,15 +32,24 @@ class UriSafeTokenGenerator implements TokenGeneratorInterface
...
+     *                                                generating entropy
+     * @param int                            $entropy The amount of entropy collected for
+     *                                                each token (in bits)
      */
-    public function __construct(SecureRandomInterface $random = null, $entropy = 256)
+    public function __construct($random = null, $entropy = 256)
```
Comment (Tobion) [inline src/Symfony/Component/Security/Csrf/TokenGenerator/UriSafeTokenGenerator.php]: $random should be removed, for additional params use func_num_args

**Comment by Tobion** [comment]
:+1: apart from the above comments

**Comment by fabpot** [comment]
should also be rebased to resolve conflict.

**Comment by pierredup** [comment]
Done

**Comment by fabpot** [comment]
Thank you @pierredup.

**Pull request merged**